File: python/covertable/sorters/hash.py
"""Hash sorter
"""


def fnv1a32(s):
    h = 0x811C9DC5
    for c in s.encode("utf-8"):
        h ^= c
        h = (h * 0x01000193) & 0xFFFFFFFF
    return format(h, '08x')


def sort(pairs, salt="", indices=None, **kwargs):
    def comparer(pair):
        if indices:
            key = "{} {}".format(",".join(str(indices[n]) for n in pair), salt)
        else:
            key = "{} {}".format(",".join(str(n) for n in pair), salt)
        return fnv1a32(key)

    return sorted(pairs, key=comparer)

File: typescript/src/sorters/hash.ts
import {fnv1a32} from '../lib';
import type {
  PairType,
  SortArgsType,
} from '../types';

export default function (
  pairs: PairType[],
  sortArgs: SortArgsType,
): PairType[] {
  const {salt, indices} = sortArgs;
  const comparer = (a: PairType, b: PairType) => {
    const aKey = `${a.map((n) => indices.get(n) as number)} ${salt}`;
    const bKey = `${b.map((n) => indices.get(n) as number)} ${salt}`;
    return fnv1a32(aKey) > fnv1a32(bKey) ? 1 : -1;
  }
  return pairs.sort(comparer);
};
